File: src/main/java/org/ldk/structs/Option_TxOutZ.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;


/**
 * An enum which can either contain a crate::c_types::TxOut or not
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class Option_TxOutZ extends CommonBase {
	private Option_TxOutZ(Object _dummy, long ptr) { super(ptr); }
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		super.finalize();
		if (ptr != 0) { bindings.COption_TxOutZ_free(ptr); }
	}
	static Option_TxOutZ constr_from_ptr(long ptr) {
		bindings.LDKCOption_TxOutZ raw_val = bindings.LDKCOption_TxOutZ_ref_from_ptr(ptr);
		if (raw_val.getClass() == bindings.LDKCOption_TxOutZ.Some.class) {
			return new Some(ptr, (bindings.LDKCOption_TxOutZ.Some)raw_val);
		}
		if (raw_val.getClass() == bindings.LDKCOption_TxOutZ.None.class) {
			return new None(ptr, (bindings.LDKCOption_TxOutZ.None)raw_val);
		}
		assert false; return null; // Unreachable without extending the (internal) bindings interface
	}

	/**
	 * When we're in this state, this COption_TxOutZ contains a crate::c_types::TxOut
	 */
	public final static class Some extends Option_TxOutZ {
		public final org.ldk.structs.TxOut some;
		private Some(long ptr, bindings.LDKCOption_TxOutZ.Some obj) {
			super(null, ptr);
			long some = obj.some;
			TxOut some_conv = new TxOut(null, some);
			this.some = some_conv;
		}
	}
	/**
	 * When we're in this state, this COption_TxOutZ contains nothing
	 */
	public final static class None extends Option_TxOutZ {
		private None(long ptr, bindings.LDKCOption_TxOutZ.None obj) {
			super(null, ptr);
		}
	}
	/**
	 * Constructs a new COption_TxOutZ containing a crate::c_types::TxOut
	 */
	public static Option_TxOutZ some(org.ldk.structs.TxOut o) {
		long ret = bindings.COption_TxOutZ_some(o.ptr);
		Reference.reachabilityFence(o);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Constructs a new COption_TxOutZ containing nothing
	 */
	public static Option_TxOutZ none() {
		long ret = bindings.COption_TxOutZ_none();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	long clone_ptr() {
		long ret = bindings.COption_TxOutZ_clone_ptr(this.ptr);
		Reference.reachabilityFence(this);
		return ret;
	}

	/**
	 * Creates a new COption_TxOutZ which has the same data as `orig`
	 * but with all dynamically-allocated buffers duplicated in new buffers.
	 */
	public Option_TxOutZ clone() {
		long ret = bindings.COption_TxOutZ_clone(this.ptr);
		Reference.reachabilityFence(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(this); };
		return ret_hu_conv;
	}

}

File: c_sharp/src/org/ldk/structs/Option_TxOutZ.cs
using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {

/**
 * An enum which can either contain a crate::c_types::TxOut or not
 */
public class Option_TxOutZ : CommonBase {
	protected Option_TxOutZ(object _dummy, long ptr) : base(ptr) { }
	~Option_TxOutZ() {
		if (ptr != 0) { bindings.COption_TxOutZ_free(ptr); }
	}

	internal static Option_TxOutZ constr_from_ptr(long ptr) {
		long raw_ty = bindings.LDKCOption_TxOutZ_ty_from_ptr(ptr);
		switch (raw_ty) {
			case 0: return new Option_TxOutZ_Some(ptr);
			case 1: return new Option_TxOutZ_None(ptr);
			default:
				throw new ArgumentException("Impossible enum variant");
		}
	}

	/** A Option_TxOutZ of type Some */
	public class Option_TxOutZ_Some : Option_TxOutZ {
		public TxOut some;
		internal Option_TxOutZ_Some(long ptr) : base(null, ptr) {
			long some = bindings.LDKCOption_TxOutZ_Some_get_some(ptr);
			TxOut some_conv = new TxOut(null, some);
			this.some = some_conv;
		}
	}
	/** A Option_TxOutZ of type None */
	public class Option_TxOutZ_None : Option_TxOutZ {
		internal Option_TxOutZ_None(long ptr) : base(null, ptr) {
		}
	}
	/**
	 * Constructs a new COption_TxOutZ containing a crate::c_types::TxOut
	 */
	public static Option_TxOutZ some(org.ldk.structs.TxOut o) {
		long ret = bindings.COption_TxOutZ_some(o.ptr);
		GC.KeepAlive(o);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Constructs a new COption_TxOutZ containing nothing
	 */
	public static Option_TxOutZ none() {
		long ret = bindings.COption_TxOutZ_none();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	internal long clone_ptr() {
		long ret = bindings.COption_TxOutZ_clone_ptr(this.ptr);
		GC.KeepAlive(this);
		return ret;
	}

	/**
	 * Creates a new COption_TxOutZ which has the same data as `orig`
	 * but with all dynamically-allocated buffers duplicated in new buffers.
	 */
	public Option_TxOutZ clone() {
		long ret = bindings.COption_TxOutZ_clone(this.ptr);
		GC.KeepAlive(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_TxOutZ ret_hu_conv = org.ldk.structs.Option_TxOutZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(this); };
		return ret_hu_conv;
	}

}
} } }
